### PR TITLE
Build: Simplify Dashboard Build Toolchain

### PR DESCRIPTION
**Pull request by ockham**
WIP. Prep work for #12072.

One core thing I've identified is that `tools/builder/react.js` uses `static.jsx` to produce minimal static markup (HTML) versions of the dashboard (really just "Turn on your JavaScript" notices). It `require()`s the webpack-bundled version of `static.jsx` to that end, and attaches `renderToStaticMarkup()`-produced strings to `window`, which it then writes to HTML files.

Getting rid of the complexity added by passing information by attaching to `window` (and thus requiring `jsdom` etc) is the main objective of this PR.

Maybe we can instead produce those HTML files as by-products of the bundling, i.e. through a Webpack plugin or loader? https://github.com/GoogleChromeLabs/prerender-loader#string-prerendering looks very promising!

Alternatively, we shouldn't bundle but just transpile `static.jsx` so we can simply `require()` it from `tools/builder/react.js`.

#### Changes proposed in this Pull Request:

Don't use `window`/`jsdom` in build system.

#### Testing instructions:

```
yarn distclean && yarn
yarn build-client
yarn docker:up
```

- Navigate to `/wp-admin/admin.php?page=jetpack_modules`
- Disable JavaScript
- See this message?

image

#### Proposed changelog entry for your changes:

Build: Simplify Dashboard Build Toolchain

### CODE REVIEW

File: _inc/client/state/initial-state/actions.js [hunk 1]
```diff
@@ -3,11 +3,7 @@
...
-	return dispatch => {
-		dispatch( {
-			type: JETPACK_SET_INITIAL_STATE,
-			initialState: window.Initial_State,
-		} );
-	};
-};
+export const setInitialState = () => ( {
+	type: JETPACK_SET_INITIAL_STATE,
+	initialState: window.Initial_State,
```
Comment (ockham) [inline _inc/client/state/initial-state/actions.js]: We should probably parametrize the action for `initialState` to put the burden of instantiating it with `window.Initial_State` on the consumer.

File: _inc/client/static.jsx [hunk 2]
```diff
@@ -12,23 +12,24 @@ import store from 'state/redux-store';
...
-	</div>
-);
+export default () =>
+	renderToStaticMarkup(
+		<div>
+			<Provider store={ store }>
+				<StaticMain />
+			</Provider>
+		</div>
+	);
```
Comment (ockham) [inline _inc/client/static.jsx]: See https://github.com/GoogleChromeLabs/prerender-loader#string-prerendering

File: _inc/client/static.jsx [hunk 3]
```diff
@@ -12,23 +12,24 @@ import store from 'state/redux-store';
...
+window.noscriptNotice = renderToStaticMarkup(
 	<Provider store={ store }>
 		<noscript>
 			<StaticWarning />
 		</noscript>
 	</Provider>
 );
 
-window.versionNotice = Server.renderToStaticMarkup(
+window.versionNotice = renderToStaticMarkup(
```
Comment (ockham) [inline _inc/client/static.jsx]: Will probably need to move `noscriptNotice` and `versionNotice` to separate files,

File: tools/builder/react.js [hunk 4]
```diff
@@ -52,7 +48,7 @@ gulp.task( 'react:master', function( done ) {
 				return;
 			}
 
-			buildStatic( done );
+			done();
```
Comment (simison) [inline tools/builder/react.js]: Prolly just `onBuild.bind( this, done )` would work here but good also as-is.

**Comment by simison** [comment]
Rebased

File: tools/builder/react.js (hunk 4 above)
Comment (ockham) [reply tools/builder/react.js]: 3a2ec1bba

**Comment by ockham** [comment]
Closing in favor of #12381.

**Pull request closed without merging**